### ISSUE TITLE
%{key} not being substituted?

### ISSUE THREAD

**Issue by traylenator** [crayfishx/hiera-http#46]
Hi,
I'm trying to query this api -https://home-assistant.io/developers/rest_api/

```
# curl -s -X GET -H "Content-Type: application/json" 'http://ha.ash.lan:8123/api/states/switch.block_facebook' | json_pp 
{
   "attributes" : {
      "friendly_name" : "Block Facebook Domains"
   },
   "entity_id" : "switch.block_facebook",
   "state" : "on"
}
```

Since I need to look up the state for an entity_id I was thinking I should load
the entity_id raw and then process the hash in puppet.

```yaml
---
:backends:
- yaml
- http

:hierarchy:
 - "hieradata/%{calling_class}"
  - hieradata/common

:yaml:
  :datadir: "/etc/puppet/environments/%{environment}"

:http:
  :host: ha.ash.lan
  :port: 8123
  :output: plain
  :cache_timeout: 20
  :paths:
  - "/api/states/%{key}"

```

And subsequent puppet.

```
$hash =  parse_json(hiera('switch.block_facebook'))
```

puppetmaster logs always contain:

```
Mar 15 20:56:43 puppet.ash.lan puppet-master[24746]: Undefined variable 'key';
Mar 15 20:56:43 puppet.ash.lan puppet-master[24746]: (file & line not available)
```

From the hiera command line everything is good:

```
# hiera -c /etc/puppet/hiera.yaml -d switch.block_facebook
DEBUG: 2017-03-15 21:01:25 +0000: Hiera YAML backend starting
DEBUG: 2017-03-15 21:01:25 +0000: Looking up switch.block_facebook in YAML backend
DEBUG: 2017-03-15 21:01:25 +0000: Looking for data source hieradata/common
DEBUG: 2017-03-15 21:01:25 +0000: Cannot find datafile /etc/puppet/environments/hieradata/common.yaml, skipping
DEBUG: 2017-03-15 21:01:25 +0000: [hiera-http]: Lookup switch.block_facebook from ha.ash.lan:8123/api/states/switch.block_facebook
DEBUG: 2017-03-15 21:01:25 +0000: [hiera-http]: Query returned data, parsing response as plain
{"attributes": {"friendly_name": "Block Facebook Domains"}, "entity_id": "switch.block_facebook", "last_changed": "2017-03-15T20:27:46.478535+00:00", "last_updated": "2017-03-15T20:27:46.478535+00:00", "state": "on"}
```

If I hardcode the %{key} value in the http paths then thing not working fails as below.
hiera command line still works.

```
puppet-master[26230]: value of hiera lookup_options must be a hash
puppet-master[26230]: value of hiera lookup_options must be a hash on node puppet.ash.lan

```
It does seem to just be the %{key} substitution that is not behaving itself.

If you have any suggestions , I've tried many close permutations of the above.

**Comment by crayfishx** [comment]
Hmmm this definately used to work - the code responsible for making sure `%{key}` is interpolated is here... https://github.com/crayfishx/hiera-http/blob/master/lib/hiera/backend/http_backend.rb#L59

Maybe there has been a change to the `Backend.parse_string` method - are you able to test this on an earlier version of Hiera?  (1.x)

**Comment by traylenator** [comment]
It is hiera 1.3.4. it is a pretty new puppet at 4.8.2. I'll look at the code you mention.

**Comment by traylenator** [comment]
All looking a bit with a hard coded key and failure on lookup_options.

lookup_options is looked up in http backend and that fails.

Thinking I actually need a much newer hiera to match puppet 4.8.

```
Mar 16 20:16:22 puppet.ash.lan puppet-master[8506]: Performing a hiera indirector lookup of lookup_options with options {:variables=>Scope(Class[Role]), :merge=>#<Puppet::Pops::HashMergeStrategy:0x18c0948 @options_t=#<Puppet::Pops::Types::PStructType:0x18ba240 @elements=[#<Puppet::Pops::Types::PStructElement:0x18ba258 @key_type=#<Puppet::Pops::Types::POptionalType:0x18ba270 @type=#<Puppet::Pops::Types::PStringType:0x18ba2a0 @size_type=nil, @values=["strategy"]>>, @value_type=#<Puppet::Pops::Types::POptionalType:0x18ba360 @type=#<Puppet::Pops::Types::PPatternType:0x18ba390 @patterns=[#<Puppet::Pops::Types::PRegexpType:0x18ba438 @pattern="hash", @regexp=/hash/>]>>>]>, @options={}>}
Mar 16 20:16:22 puppet.ash.lan puppet-master[8506]: hiera(): Hiera YAML backend starting
Mar 16 20:16:22 puppet.ash.lan puppet-master[8506]: hiera(): Looking up lookup_options in YAML backend
Mar 16 20:16:22 puppet.ash.lan puppet-master[8506]: hiera(): Looking for data source hieradata/role
Mar 16 20:16:22 puppet.ash.lan puppet-master[8506]: hiera(): Looking for data source hieradata/common
Mar 16 20:16:22 puppet.ash.lan puppet-master[8506]: hiera(): [hiera-http]: Lookup lookup_options from ha.ash.lan:8123/api/states/switch.block_facebook
Mar 16 20:16:22 puppet.ash.lan puppet-master[8506]: hiera(): [hiera-http]: Query returned data, parsing response as plain
Mar 16 20:16:22 puppet.ash.lan puppet-master[8506]: value of role lookup_options must be a hash
```

**Comment by fvoges** [comment]
@crayfishx This is also affecting the latest version. I've implemented a fix/workaround in #53

**Comment by crayfishx** [comment]
Fixed in 3.1.0 - thanks @fvoges
